modules/biobank/php/specimendao.class.inc

```diff
@@ -0,0 +1,719 @@
+        // with respective data.
+        if (isset($data['biobank_specimen_preparation']['SpecimenProtocolID'])) {
+            $this->db->unsafeInsertOnDuplicateUpdate(
+                'biobank_specimen_preparation',
+                $data['biobank_specimen_preparation']
+            );
+        }
+
+        // If method exists, insert or update biobank_specimen_analysis
+        // with respective data.
+        if (isset($data['biobank_specimen_analysis']['SpecimenMethodID'])) {
+            $this->db->unsafeInsertOnDuplicateUpdate(
+                'biobank_specimen_analysis',
+                $data['biobank_specimen_analysis']
+            );
+        }
+
+        // If parent exists, insert or update biobank_specimen_parent
+        // with respective data.
+        if (isset($data['biobank_specimen_parent'])) {
+            foreach ($data['biobank_specimen_parent'] as $insert) {
+                $this->db->insertOnDuplicateUpdate(
+                    'biobank_specimen_parent',
+                    $insert
+                );
+            }
+        }
+
+        // If F/T Cycle exists, insert or update biobank_specimen_freezethaw
+        // with respective data.
+        $fTCycle = $specimen->getFTCycle();
+        if (isset($fTCycle)) {
+            $this->db->insertOnDuplicateUpdate(
+                'biobank_specimen_freezethaw',
+                $data['biobank_specimen_freezethaw']
+            );
+        }
+    }
+
+    /**
+     * This function takes a Specimen array and prepares the data to be
+     * inserted into the database by converting it to a data array.
+     *
+     * @param array $specimen Specimen array to be converted
+     *
+     * @return array $data Array containing the data to be inserted
+     */
+    private function getSQLFromArray(array $specimen) : array
+    {
+        $specimenData = array(
+                         'SpecimenID'     => $specimen['id'],
+                         'ContainerID'    => $specimen['containerId'],
+                         'SpecimenTypeID' => $specimen['typeId'],
+                         'Quantity'       => $specimen['quantity'],
+                         'UnitID'         => $specimen['unitId'],
+                         'SessionID'      => $specimen['sessionId'],
+                        );
+
+        $freezeThawData = array(
+                           'SpecimenID'      => $specimen['id'],
+                           'FreezeThawCycle' => $specimen['fTCycle'],
+                          );
+
+        $collection     = $specimen->getCollection();
+        $collectionData = array(
+                           'SpecimenID' => $specimen['id'],
+                           'Quantity'   => $collection['quantity'],
+                           'UnitID'     => $collection['untiId'],
+                           'CenterID'   => $collection['centerId'],
+                           'Date'       => (new \DateTime($collection['date']))->format('Y-m-d H:i'),
```

Comment:
It may be best to write a function `toMysqlDateFormat(string $date): string` to make these calls a little cleaner. You could put it in `\Utility` since this is likely to come up in other places.

Reply:
sounds good!